actionpack/lib/action_dispatch/routing/route_set.rb

```diff
@@ -537,6 +598,61 @@ def add_route(mapping, path_ast, name, anchor)
         route
       end
 
+      def add_polymorphic_mapping(options, &block)
+        defaults = options.dup
+        klass = defaults.delete(:class)
+        if klass.nil?
+          raise ArgumentError, "Missing :class key from polymorphic mapping options"
+        end
+
+        @polymorphic_mappings[klass] = DirectUrlHelper.new(klass, defaults, &block)
+      end
+
+      def add_url_helper(name, options, &block)
+        named_routes.add_url_helper(name, options, &block)
+      end
+
+      class DirectUrlHelper
+        attr_reader :name, :defaults, :block
+
+        def initialize(name, defaults, &block)
+          @name = name
+          @defaults = defaults
+          @block = block
+        end
+
+        def call(t, args, options, outer_options = {})
+          url_options = eval_block(t, args, options)
+
+          case url_options
```

Comment:
Can we push this whole `case` back into `t` (that's the "real" route thingy, right?), by just adding a new `url_for`-like method that takes an explicit `path_only` parameter?